src/test/java/org/sagebionetworks/bridge/udd/synapse/SynapseDownloadFromTableTaskTest.java

```diff
@@ -0,0 +1,438 @@
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertNull;
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
+
+import java.io.File;
+import java.io.Reader;
+import java.io.Writer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import com.google.common.io.CharStreams;
+import org.joda.time.LocalDate;
+import org.mockito.ArgumentCaptor;
+import org.sagebionetworks.client.exceptions.SynapseException;
+import org.sagebionetworks.repo.model.file.BulkFileDownloadResponse;
+import org.sagebionetworks.repo.model.file.FileDownloadSummary;
+import org.sagebionetworks.util.csv.CsvNullReader;
+import org.testng.annotations.Test;
+
+import org.sagebionetworks.bridge.udd.dynamodb.UploadSchema;
+import org.sagebionetworks.bridge.udd.dynamodb.UploadSchemaKey;
+import org.sagebionetworks.bridge.udd.exceptions.AsyncTaskExecutionException;
+import org.sagebionetworks.bridge.udd.helper.MockFileHelper;
+
+@SuppressWarnings({ "rawtypes", "unchecked" })
+public class SynapseDownloadFromTableTaskTest {
+    private static final UploadSchemaKey TEST_SCHEMA_KEY = new UploadSchemaKey.Builder().withStudyId("test-study")
+            .withSchemaId("test-schema").withRevision(42).build();
+
+    // The default test schema should include at least 2 file handle IDs and a mix of file handles and
+    // non-file-handles.
+    private static final UploadSchema DEFAULT_TEST_SCHEMA = new UploadSchema.Builder().withKey(TEST_SCHEMA_KEY)
+            .addField("foo", "INT").addField("bar", "ATTACHMENT_BLOB").addField("baz", "ATTACHMENT_JSON_BLOB").build();
+
+    private MockFileHelper mockFileHelper;
+    private ArgumentCaptor<String> synapseQueryCaptor;
+    private ArgumentCaptor<Set> synapseFileHandleIdSetCaptor;
+    private SynapseDownloadFromTableTask task;
+    private File tmpDir;
+
+    @Test
+    public void csvHasNoUserRows() throws Exception {
+        // setup
+        String csvContent = "\"recordId\",\"healthCode\",\"foo\",\"bar\",\"baz\"";
+        setupTestWithArgs(DEFAULT_TEST_SCHEMA, csvContent, null, null);
+
+        // execute and validate
+        SynapseDownloadFromTableResult result = task.call();
+        assertNull(result.getCsvFile());
+        assertNull(result.getBulkDownloadFile());
+        postValidation(result);
+    }
+
+    @Test
+    public void csvHasNoHealthCode() throws Exception {
+        // setup
+        String csvContent = "\"recordId\",\"foo\",\"bar\",\"baz\"\n" +
+                "\"record-1\",\"37\",\"file-handle-1\",\"file-handle-2\"";
+        setupTestWithArgs(DEFAULT_TEST_SCHEMA, csvContent, null, null);
+
+        // execute
+        Exception thrownEx = null;
+        try {
+            task.call();
+            fail("expected exception");
+        } catch (AsyncTaskExecutionException ex) {
+            thrownEx = ex;
```

Comment:
As mentioned elsewhere, I don't like string matching exception messages, since this tightly couples our implementation to a specific error message. In most (all) of these cases, the interesting thing isn't the error message, but the exception type and how our code handles the exception.